### ISSUE TITLE
Frozen deps does not check for workflow pinning

### ISSUE THREAD

**Issue by laurentsimon** [ossf/scorecard#412]
In our own repo:
```
$grep uses .github/workflows/* | grep master
.github/workflows/codescan.yml:        uses: aquasecurity/trivy-action@master
.github/workflows/integration.yml:        uses: google-github-actions/setup-gcloud@master
.github/workflows/synk-cron.yaml:        uses: snyk/actions/iac@master
.github/workflows/synk.yaml:        uses: snyk/actions/iac@master
```

Shall we pin these? If yes, we should also add this to the Frozen-Dep check.

**Comment by laurentsimon** [comment]
@naveensrinivasan to add the check in scorecard, my current plan is to download the tarball and use a regex for files under `.github/workflow/` to detect `uses: somename@` followed by `vX.Y.Z` or `X.Y.Z` where Y and Z are optional. Basically `@v?\d(\.\d){1,2}$`. I don't like that we have to download the tarball again (it's already done in another check). 

Long-term, we can probably download the tarball once and pass it to other checks that need it.

 wdut?

**Comment by naveensrinivasan** [comment]
@laurentsimon  Tags can be moved. That is the reason teams prefer to pin the SHA instead of tags. So doing checks for tags is not recommended IMO.

Scorecard avoids maintaining the state within the CLI. Passing the downloaded tarball would do that. We would like to keep it stateless. 

I agree to download the tarball again is not utilizing the resources well. But there is a bigger discussion on scaling these API calls and we will be addressing that soon.

**Comment by laurentsimon** [comment]
> @laurentsimon Tags can be moved. That is the reason teams prefer to pin the SHA instead of tags. So doing checks for tags is not recommended IMO.

yes, but do you know if hash pins are supported in github workflows?
We have a dep-pinning check for package management (go, python, etc). Is the reasoning different there? 

> 
> Scorecard avoids maintaining the state within the CLI. Passing the downloaded tarball would do that. We would like to keep it stateless.
> 
> I agree to download the tarball again is not utilizing the resources well. But there is a bigger discussion on scaling these API calls and we will be addressing that soon.

ok.

**Comment by naveensrinivasan** [comment]
> > @laurentsimon Tags can be moved. That is the reason teams prefer to pin the SHA instead of tags. So doing checks for tags is not recommended IMO.
> 
> yes, but do you know if hash pins are supported in github workflows?
> We have a version-pinning check with no hash for package management (go, python, etc). Is the reasoning different there?
> 

Yes, has pinning in supported in GitHub workflows

https://julienrenaux.fr/2019/12/20/github-actions-security-risk/#solution-use-commit-hash-as-version

> > Scorecard avoids maintaining the state within the CLI. Passing the downloaded tarball would do that. We would like to keep it stateless.
> > I agree to download the tarball again is not utilizing the resources well. But there is a bigger discussion on scaling these API calls and we will be addressing that soon.
> 
> ok.

**Comment by laurentsimon** [comment]
Brilliant, thanks for the link. I've updated #417 accordingly.
To clarify, shall I go ahead and add the check to scorecard for hash-pinned deps?
Shall I first pin all deps by hash? I'm kinda worried we're never going to think of updating them. Does dependabot support github workflows and will it alert us when new versions are released?

**Comment by naveensrinivasan** [comment]
> Brilliant, thanks for the link. I've updated #417 accordingly.
> To clarify, shall I go ahead and add the check to scorecard for hash-pinned deps?

Ideally, we would need to have the actions YAML parsed into a struct and then look for SHA. String matching is error-prone and would give false positives.  Also, we would have to check if YAML's have a standard structure to be parsed into a Go struct.

> Shall I first pin all deps by hash? I'm kinda worried we're never going to think of updating them. Does dependabot support github workflows and will it alert us when new versions are released?

**Comment by inferno-chromium** [comment]
> > Brilliant, thanks for the link. I've updated #417 accordingly.
> > To clarify, shall I go ahead and add the check to scorecard for hash-pinned deps?
> 
> Ideally, we would need to have the actions YAML parsed into a struct and then look for SHA. String matching is error-prone and would give false positives. Also, we would have to check if YAML's have a standard structure to be parsed into a Go struct.

+1, we should just use a yaml lib to parse these. 
we should ideally support 1-2 CI/CD systems, e.g. github actions, cloudbuild, both of them use yaml
for cloudbuild, see this example - https://github.com/google/clusterfuzz/blob/master/cloudbuild.yaml#L16
"name: gcr.io/cloud-builders/bazel@sha256:7360c36bded15db68a35cfb1740a994f0a09ad5ce378a97f96d698bc223e442a"



> 
> > Shall I first pin all deps by hash? I'm kinda worried we're never going to think of updating them. Does dependabot support github workflows and will it alert us when new versions are released?

**Comment by naveensrinivasan** [comment]
Just an FYI the GitHub YAML can be hard to parse as schema definition could have lots of definitions here is a good starting point https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsuses

**Comment by laurentsimon** [comment]
> > Brilliant, thanks for the link. I've updated #417 accordingly.
> > To clarify, shall I go ahead and add the check to scorecard for hash-pinned deps?
> 
> Ideally, we would need to have the actions YAML parsed into a struct and then look for SHA. String matching is error-prone and would give false positives. Also, we would have to check if YAML's have a standard structure to be parsed into a Go struct.

That would be ideal, yeah. We could also start with the following regex `^\s*-*\s*uses:\s*[a-zA-Z\/]*@[\da-z]{40,}\s*$`
It matches `uses` and dependency on the same line, which I think is what most people use. 

Let me know if you think we should build it incrementally or at once. I don't mind either.

Thanks both of you for the links!

> > Shall I first pin all deps by hash? I'm kinda worried we're never going to think of updating them. Does dependabot support github workflows and will it alert us when new versions are released?

**Comment by laurentsimon** [comment]
> What if the yaml has some other SHA which is not the actual SHA of the workflow step?

Note sure what you mean. Link?

> What if the yaml uses more than one `step` and only one of them is using SHA?

that would be counted as bad and flagged.

> With these, it will be returning false-positive results. We should not be returning false positives. That is the reason string parsing is error-prone. We should parse the YAML struct and validate it for SHA.

**Comment by laurentsimon** [comment]
Fixed by #466
